uint32 center_x
uint32 center_y
uint32 rad
uint32[] im